libnavigation-core/src/main/java/com/mapbox/navigation/core/fasterroute/FasterRouteDetector.kt

```diff
@@ -3,11 +3,15 @@ package com.mapbox.navigation.core.fasterroute
 import com.mapbox.api.directions.v5.models.DirectionsRoute
 import com.mapbox.navigation.base.trip.model.RouteProgress
 
-internal class FasterRouteDetector {
-    fun isRouteFaster(newRoute: DirectionsRoute, routeProgress: RouteProgress): Boolean {
-        val newRouteDuration = newRoute.duration() ?: return false
+internal class FasterRouteDetector(
+    private val routeComparator: RouteComparator
+) {
+
+    fun isRouteFaster(alternativeRoute: DirectionsRoute, routeProgress: RouteProgress): Boolean {
+        val alternativeDuration = alternativeRoute.duration() ?: return false
         val weightedDuration = routeProgress.durationRemaining * PERCENTAGE_THRESHOLD
-        return newRouteDuration < weightedDuration
+        val isNewRouteFaster = alternativeDuration < weightedDuration
+        return isNewRouteFaster && routeComparator.isNewRoute(routeProgress, alternativeRoute)
```

Comment:
I believe `routeComparator.isNewRoute(routeProgress, alternativeRoute)` is going to be `true` most of the times as `alternativeRoute` is going to have a different `geometry` most likely and we're always passing the `alternativeRoute` 👀 `FasterRouteController`

```kotlin
private val fasterRouteRequestCallback = object : RoutesRequestCallback {
    override fun onRoutesReady(routes: List<DirectionsRoute>) {
        val currentRoute = directionsSession.routes.firstOrNull()
            ?: return
        tripSession.getRouteProgress()?.let { progress ->
            val isAlternativeFaster = fasterRouteDetector.isRouteFaster(routes[0], progress)
            fasterRouteObserver?.onFasterRoute(currentRoute, routes, isAlternativeFaster)
        }
    }
```

Reply:
Yeah you're totally right. Pushed the breaking test case

Also turns out the legacy approach has the same issue. It can only compare routes with the same origin and destination. After completing many steps, the the summary of the route changes.

Here is legacy code that identifies a route https://github.com/mapbox/mapbox-navigation-android/blob/2476bab1fd0201af45e43c33dedbc20d991e1958/libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/RouteComparator.java#L58-L65

When testing a route, all of these are on the same route:
Route start: "Monterey Boulevard, Taraval Street"
Mid route: "Taraval Street, 28th Avenue"
End of route: "28th Avenue"

If faster route happens at any of these stages, it would think they are different routes